File: 22-oop-in-javascript/employee.rb
class Employee
  def initialize
    self.name = ''
    self.dept = 'general'
  end
end

# new_employee = Employee.new


class Manager < Employee
  def initialize
    super
    self.reports = []
  end
end


class WorkerBee < Employee
  def initialize
    super
    self.projects = []
  end
end



class Sales < WorkerBee
  def initialize
    super
    self.dept = 'sales'
    self.quota = 100
  end
end

File: 22-oop-in-javascript/employee.js
function Employee() {
  this.name = '';
  this.dept = 'general';
}



// newEmployee = new Employee();


function Manager() {
  // `this` is an object of type Manager
  Employee.call(this);
  this.reports = [];
}

Manager.prototype = Object.create(Employee.prototype);

function WorkerBee() {
  // `this` is an object of type Manager
  Employee.call(this);
  this.projects = [];
}

WorkerBee.prototype = Object.create(Employee.prototype);


function Sales() {
  WorkerBee.call(this);
  this.dept = 'sales';
  this.quote = 100;
  // this.hello = function() {
  //
  // }
}

Sales.prototype = Object.create(WorkerBee.prototype);

// ES2015
class Sales extends WorkerBee {
  constructor {
    this.dept = 'sales';
    this.quote = 100;
  }
}


// Adding a behaviour to an object
Sales.prototype.hello = function() {
  return "Hello from " + this.name;
}
